src/libs/elektra/opmphmpredictor.c

```diff
@@ -0,0 +1,203 @@
+	/*
+	 * update patterTable with worth information
+	 */
+	// check if previos lookup sequence was worth hashing
+	size_t worthOpmphm = opmphmPredictorWorthOpmphm (n);
+	uint8_t wasItWorth = op->lookupCount > worthOpmphm ? 1 : 0;
+	// find position in array
+	uint16_t pos = op->history & OPMPHM_PREDICTOR_HISTORY_EXTRACTION_MASK;
+	uint8_t * state = &(op->patternTable[pos >> 2]); // (pos * 2) / 8 == pos >> 2
+	// extract 2 bit state
+	uint8_t newState = (*state >> ((pos & 0x3) << 1)) & 0x3; // ((pos & 0x3) << 1) == (pos % 4) * 2), 0x3 is 2 bit mask
+	// feed state and worth information to the predictionAutomata
+	newState = predictionAutomata[newState][wasItWorth];
+	// delete old state
+	*state &= ~(0x3 << ((pos & 0x3) << 1)); // ((pos & 0x3) << 1) == (pos % 4) * 2), 0x3 is 2 bit mask
+	// store the new state
+	*state |= newState << ((pos & 0x3) << 1); // ((pos & 0x3) << 1) == (pos % 4) * 2), 0x3 is 2 bit mask
+	/*
+	 * predict with updated history
+	 */
+	// add the worth information to the history
+	op->history = ((op->history << 1) | wasItWorth) & OPMPHM_PREDICTOR_HISTORY_EXTRACTION_MASK;
+	pos = op->history & OPMPHM_PREDICTOR_HISTORY_EXTRACTION_MASK;
+	state = &(op->patternTable[pos >> 2]); // (pos * 2) / 8 == pos >> 2
+	// extract 2 bit state
+	uint8_t prediction = (*state >> ((pos & 0x3) << 1)) & 0x3; // ((pos & 0x3) << 1) == (pos % 4) * 2), 0x3 is 2 bit mask
+	// reset lookupCount
+	op->lookupCount = 1;
+	// determine prediction
+	if (prediction > 1)
+	{
+		return 1;
+	}
+	else
+	{
+		return 0;
+	}
+}
+
+/**
+ * @brief Allocates and initializes the OpmphmPredictor.
+ *
+ * Reserves for all possible values of OPMPHM_PREDICTOR_HISTORY_EXTRACTION_MASK two bits
+ * to store all 4 states of the Prediction Automata A2.
+ * Sets the initial state to 0.
+ *
+ * @retval OpmphmPredictor * success
+ * @retval NULL memory error
+ */
+OpmphmPredictor * opmphmPredictorNew (void)
+{
+	OpmphmPredictor * out = elektraCalloc (sizeof (OpmphmPredictor));
+	if (!out)
+	{
+		return NULL;
+	}
+	uint16_t bytesInPatternTable = (OPMPHM_PREDICTOR_HISTORY_EXTRACTION_MASK + 1) >> 2; // x >> 2 == (x * 2) / 8
+	if (!bytesInPatternTable)
+	{
+		bytesInPatternTable = 1;
+	}
+	out->patternTable = elektraCalloc (bytesInPatternTable * sizeof (uint8_t));
+	//~ memset (out->patternTable, 0x0, bytesInPatternTable * sizeof (uint8_t));
+	if (!out->patternTable)
+	{
+		elektraFree (out);
+		return NULL;
+	}
+	out->size = bytesInPatternTable;
+	return out;
+}
+
+/**
+ * @brief Make a copy of the OpmphmPredictor.
+ *
+ * @param source the OpmphmPredictor source
+ * @param dest the OpmphmPredictor destination
+ */
+void opmphmPredictorCopy (OpmphmPredictor * dest, OpmphmPredictor * source)
+{
+	ELEKTRA_NOT_NULL (source);
+	ELEKTRA_NOT_NULL (dest);
+	// copy values
+	dest->history = source->history;
+	dest->size = source->size;
+	dest->lookupCount = source->lookupCount;
+	// copy patternTable
+	memcpy (dest->patternTable, source->patternTable, source->size * sizeof (uint8_t));
```

Comment:
there is a elektraMemcpy

Reply:
elektraMemcpy takes a Key as argument and the size is the per Key size:
`ssize_t elektraMemcpy (Key ** array1, Key ** array2, size_t size);`
I think it is not applicable for any `void *`.

Reply:
Yes, you are right.